Cj.bob get started (#1847)
hashicorp/web-unified-docs: content/bob/bob-tutorial/modernize-node-api.md

---
id: modernize-node-api

> name: Modernize a Node.js API with Bob

Reviewer: This should probably be "Get Started with Bob" or "Bob Quickstart"? Including the application type seems a little specific and I don't want people to think that this is the main use case for Bob.

short_name: Modernize Node.js API
products_used:
  - bob
description: >-

>   Learn to modernize a Node.js Express API from version 16 to 22 using IBM Bob. Master AI-assisted development with modes, approvals, and literate coding in this hands-on tutorial.

Reviewer: ```suggestion
  Learn to use Bob for application modernization by upgrading a Node.js Express API from version 16 to 22 using IBM Bob. Master AI-assisted development with modes, approvals, and literate coding in this hands-on tutorial.
```

---

> # Modernize a Node.js API with Bob

Reviewer: This should be whatever we decide to put in the title. Should not include the app type.

IBM Bob is an AI-powered Integrated Development Environment (IDE) and Software Development Lifecycle (SDLC) partner for developers. Bob is a standalone application that you install on your computer.

> Bob helps you write code by separating intent (your goal), evidence (what exists), and judgment (your decision). Bob works in two phases: planning and execution, ensuring you stay in control while upgrading applications, refactoring APIs, and managing dependencies.

Reviewer: Is this how Bob is different from other AI IDEs or is this common across any?

Author: I'm not sure there is another IDE that has AI integrated as a feature like Bob does. For example, if I use VSC, I have a Claude or Github plugin installed. Then the AI interacts with the text editor. 

Bob isn't a plugin, it's part of the actual text editor. 

That text is also what I pulled from how bob thinks book.

---

> id: modernize-node-api

Author: This is going in IBM docs.I don't know how it works with IBM. I'll let Liam handle it. I

Bob provides the following five key capabilities that you learn by modernizing a working API:

1. **Modes** - Bob has different modes which differ in permissions and workflows. The primary modes are Plan, Code, Ask, and Advanced.
1. **Context mentions** let you reference specific elements of your project in your conversations with Bob, such as specific file, folder, or Git commits.
1. **Approval workflow** lets you review every file edit and command before Bob executes it.
1. **Code actions** provide quick fixes, refactorings, and AI-powered suggestions in your editor.
1. **Literate coding** lets you write code with AI assistance in your editor. You type instructions in plain language right where the code should go.

## Prerequisites

This tutorial uses a TypeScript Express REST API as the example project. However, you do not need experience in Node.js or TypeScript.

To complete this tutorial, you need:

- **[IBM Bob IDE](https://bob.ibm.com/docs/ide/install)** - Download and install the IBM Bob application on your computer. Bob is a standalone IDE application and not an extension.

> - **[Docker](https://docs.docker.com/get-docker/)** - To run containerized builds. You won't need to install Node.js or dependencies locally.

Reviewer 2: I don't think this is something we have called out explicitly in our style guide, but one norm is we _usually_ use colons rather than hyphens in this case. The important thing is that it's consistent, but might be worth double checking the style guide.

## Understanding Bob's AI development modes

Bob has different modes for different tasks. Each mode has a different set of tools and features. The following are Bob's default modes and when you should use them:

- **Plan** - When you want to analyze requirements, research and design implementation steps
- **Code** - When you want Bob to make changes or run commands 
- **Ask** - When you want to explore and understand code without making changes
- **Advanced** - When you need all of Bob's tools for complex workflows 

> - **Orchestrator** - Coordinates complex tasks across multiple modes

Reviewer 2: I'm not super clear on the difference between Advanced and Orchestrator here, or what qualifies as "complex workflows"

## Open Bob and clone the repository

> 

Reviewer 2: Add a paragraph to introduce this section

1. Launch the IBM Bob application on your computer. Look for IBM Bob in your Applications folder (macOS), Start menu (Windows), or applications menu (Linux). 

1. If Bob is not already opened, open the Bob chat panel by clicking the Bob icon beside the navigation bar or use the shortcut `Cmd+L` (Mac) or `Ctrl+L` (Windows/Linux)

   ![Bob chat panel open in IBM Bob IDE](img/open-bob.png)

   The Bob chat panel has three main components:
   - **Chat history** - Displays your conversation with Bob at the top
   - **Input field** - At the bottom of the panel, type your requests to Bob using natural language
   - **Send button** - Click the paper plane icon to the right of the input field, or press **Enter** to send your message

   You interact with Bob by typing natural language requests in the input field, just like chatting with a colleague.

> 1. Clone the tutorial repository by clicking on the files icon located on the top left of the Bob IDE panel. Then click **Clone Repository**, and paste the following GitHub repository into the search bar. Bob will ask you where you want to save the repository. You can save it wherever you like:

Reviewer 2: ```suggestion
1. Clone the tutorial repository by clicking on the files icon located on the top left of the Bob IDE panel. Then click **Clone Repository**, and paste the following GitHub repository into the search bar. Bob asks you where you want to save the repository. You can save it wherever you like:
```

   The directory contains a TypeScript-based Express.js TODO API with the following structure:
   - Source code in `src/` directory
   - Database configuration (`db.ts`)
   - Main entry point (`index.ts`)
   - Models and routes for TODO management
   - Configuration files (package.json, tsconfig.json)
   - Documentation (README.md, SETUP-VERIFIED.md, TESTING-NOTES.md, MODERNIZATION.md)
   ``` 

## Create a Dockerfile

Create a `Dockerfile` in the express-todo-api folder to run the application in a containerized environment. With Docker, you don't need to install Node.js or dependencies on your machine.

1. Switch Bob to code mode by clicking the mode selector dropdown to the left of the chat input field and selecting **Code**.

   ![Mode selector dropdown showing Code mode selection](img/modes-code-dropdown.png)

1. Ask Bob to create the Dockerfile in the Bob chat panel:

   ```text
   Create a Dockerfile for this Node.js TypeScript application. It should use Node 16 to match the current application version, build the TypeScript code, and run the app on port `3000`.
   ```

   ![Bob proposes Dockerfile with Save action button](img/create-dockerfile.png)

>    Bob determines the appropriate Dockerfile structure and shows you the proposed file. Action buttons appear above the chat input field. Click **Save** to create the file. Later in this tutorial, Bob updates the Dockerfile to use Node 22.

Reviewer 3: I must have a different setting turned on, as It is asking me to read the package.json file to understand the build and start scripts.  I clicked Approve, and then I was presented with the Save button to create the new file.

## Verify the legacy application builds

Before you have Bob modernize the legacy application, verify the legacy code works. Ask Bob to build and test the legacy application using Docker.

> In the Bob chat panel, tell Bob what you want it to do:

Reviewer 2: ```suggestion
In the Bob chat panel, tell Bob to build and test the application:
```

```text
Build and test the legacy application using Docker to verify it works correctly before modernization. The application should respond to API requests at /api/todos with sample data

After testing, clean up the container.
```

Bob understands your intent and determines the right Docker commands to use. Bob shows you the commands and asks for approval. Click **Run** to execute them.

> 

Reviewer 3: Ok, so build failed, but I let Bob edit the file....

<img width="292" height="301" alt="Image" src="https://github.com/user-attachments/assets/88f4afe2-3fc6-4f1e-9e6a-e6ccbbca5978" />

Bob gives you something similar to following output to confirm the legacy application builds:

```
Successfully built, tested, and cleaned up the legacy Express TODO API application using Docker:

**Build Results:**
- Created Dockerfile using Node 16 Alpine with multi-stage build
- Successfully built Docker image with TypeScript compilation
- Image size optimized with production-only dependencies

**Test Results:**
- Container ran successfully on port `3000`
- API endpoint `/api/todos` responded correctly with sample data:
  ```json
  [{
    "id": "0133de51-d9ba-490a-93c9-e9af8bb58a33",
    "title": "Learn Node.js modernization",
    "completed": false,
    "createdAt": "2026-02-16T16:15:56.627Z"
  }]
  

**Cleanup:**
- Stopped and removed container `express-todo-api-test`
- Removed Docker image `express-todo-api`

The legacy application is verified to work correctly and is ready for modernization. The Dockerfile remains in the project for future use.

```

> You and Bob verified the legacy application is working correctly.

Reviewer 3: I had to click Run a few times as it progressed through testing.  It then asked to "clean up by stopping and removing the container" so I said yes???

## Explore Node.js codebase with Bob's Ask mode

> Before modernizing, switch to Bob's **Ask mode** in the same dropdown that you used to change Bob to Code mode. Ask mode can only read files, so you can safely explore code without accidentally making changes.

Reviewer 2: ```suggestion
Before modernizing, click the mode selector drop down and switch to Bob's **Ask mode**. Ask mode can only read files, so you can safely explore code without accidentally making changes. You can also type `/ask` in Bob chat to change to Ask mode.
```

Click the mode selector dropdown to the left of the chat input field and select **Ask** or type `/ask` in Bob chat.

> You will explore three files that you need to modernize: `package.json`, `tsconfig.json`, and `Dockerfile`, by explicitly telling Bob which files to analyze using @ mentions. You will understand the modernization requirements before making any changes.

Reviewer 2: ```suggestion
```

**Summary:** The project uses legacy CommonJS with ES2020 target. For Node.js 22, you should update the target/lib to ES2022+ and consider migrating to native ESM modules (though CommonJS still works fine).
```

### Analyze the Dockerfile

Ask Bob to analyze the project's Dockerfile:

```text
@Dockerfile What Node.js version does this Dockerfile use? What would need to change for Node.js 22? Give me a short summary.
```

Bob reads the `Dockerfile` and responds with what needs to be updated for Node.js 22.

In summary, Bob tells you:

```text
**Current Dockerfile Node.js Version:** Node 16 Alpine (`node:16-alpine`)

**Changes Needed for Node.js 22:**

1. **Base Image (Line 3):** `FROM node:16-alpine` → `FROM node:22-alpine`
2. **Production Image (Line 18):** `FROM node:16-alpine` → `FROM node:22-alpine`

**Summary:** Only 2 lines need to change - update both `FROM` statements from `node:16-alpine` to `node:22-alpine`. The multi-stage build structure, commands, and configuration remain the same since the Dockerfile is already well-structured and Node.js 22 maintains backward compatibility with the build process.
```

## Modernize the API with Bob's Code mode

> You and Bob know the modernization requirements for `@package.json`, `@tsconfig.json`, and `@Dockerfile`.

Reviewer 2: ```suggestion
Bob determined the following modernization requirements for `package.json`, `tsconfig.json`, and `Dockerfile`:

```

- Node.js 16 → 22 in package.json and Dockerfile
- TypeScript compiler target and module settings

> - Outdated dependency versions

Reviewer 2: ```suggestion
- Change Node.js 16 to Node.js 22 in package.json and Dockerfile
- Update the TypeScript compiler target and module settings
- Fix outdated dependency versions
```

Bob's **literate coding** feature lets you write code with AI assistance in your editor. Instead of using the chat panel, you write instructions in plain language right where the code should go.

Use literate coding to mark the app as modernized by updating the sample todo title:

1. In the Explorer tab on the left side of the screen, navigate to the `express-todo-api` folder and click on `src/db.ts` to open it in the editor

   ![db.ts file open in the editor](img/db.ts.png)

2. Toggle literate coding mode by clicking the magic wand icon in the editor toolbar (or press `Cmd+M` on Mac / `Ctrl+M` on Windows/Linux)

   ![Literate coding mode activated](img/literate-coding.png)

3. In the `src/db.ts` file that's now open in your editor, paste the following instruction at the top of the file, before any code:

   ```text
   // Find the sample todo object's title property and change it to "Welcome to the Modernized Node.js 22 API"
   ```

   Bob understands this comment is an instruction (not regular code) because you're in literate coding mode. Your instruction appears highlighted in the editor.

   ![Instruction highlighted in literate coding mode](img/literate-run.png)

> 4. Click **Generate** at the bottom of the document

Reviewer 2: ```suggestion
4. Click **Generate** at the bottom of the document.
```

   **Before:**
   ```typescript
   title: "Learn Node.js modernization"
   ```

   **After:**
   ```typescript
   title: "Welcome to the Modernized Node.js 22 API"
   ```

   ![Bob shows the generated changes with Accept all button](img/literate-generate.png)

5. Click **Accept all** at the bottom of the document where you clicked **Generate** to apply the change to the file

6. Click **Exit** below the literate magic wand at the top of the file to exit literate coding mode

You've updated the API using literate coding. When you test the API in the following section, you see this new title in the API response.

## Verify the modernized application with Docker

Verify the code modernization and your literate code change by rebuilding and testing the Docker container.

In Code mode, describe what you want Bob to verify:

```text
Verify the modernized Node.js 22 application works by building and testing it with Docker. Check that the API responds at /api/todos and shows the updated todo title. Clean up when done. Give me a short summary. I also want to see the validation of the new title.
```

> Once Bob verifies the application has been modernized, it gives you the following summary.

Author: So I'm not sure how to handle this. This is because of auto-approve being off. It is going to ask the user to save, approve, run a LOT.
